source/solution_deploy/lib/member/wait-provider.ts

```diff
@@ -0,0 +1,142 @@
+// SPDX-License-Identifier: Apache-2.0
+import { createHash, Hash } from 'crypto';
+import { CfnCustomResource, CustomResource, Duration, Stack } from 'aws-cdk-lib';
+import { Code, Function, Runtime } from 'aws-cdk-lib/aws-lambda';
+import { Bucket } from 'aws-cdk-lib/aws-s3';
+import { CfnDocument } from 'aws-cdk-lib/aws-ssm';
+import { Construct } from 'constructs';
+import { PolicyDocument, PolicyStatement, Role, ServicePrincipal } from 'aws-cdk-lib/aws-iam';
+import { NagSuppressions } from 'cdk-nag';
+
+export interface SsmDocumentWaitProviderProps {
+  readonly solutionDistBucket?: string;
+  readonly solutionVersion?: string;
+  readonly solutionTMN?: string;
+  readonly runtimePython?: Runtime;
+  readonly serviceToken?: string;
+}
+
+export interface SsmDocumentWaitResourceProps {
+  readonly document: CfnDocument;
+}
+
+export class SsmDocumentWaitProvider extends Construct {
+  public readonly serviceToken: string;
+
+  constructor(scope: Construct, id: string, props: SsmDocumentWaitProviderProps) {
+    super(scope, id);
+
+    if (!props.serviceToken) {
+      if (!props.solutionDistBucket || !props.solutionVersion || !props.solutionTMN || !props.runtimePython) {
+        throw new Error('If not specifying service token, you must specify all required properties');
+      }
+
+      const policyDocument = new PolicyDocument({
+        statements: [
+          new PolicyStatement({ actions: ['cloudwatch:PutMetricData'], resources: ['*'] }),
+          new PolicyStatement({
+            actions: ['logs:CreateLogGroup', 'logs:CreateLogStream', 'logs:PutLogEvents'],
+            resources: ['*'],
+          }),
+        ],
+      });
+
+      const role = new Role(this, 'Role', {
+        assumedBy: new ServicePrincipal('lambda.amazonaws.com'),
+        inlinePolicies: { LambdaPolicy: policyDocument },
+      });
+
+      NagSuppressions.addResourceSuppressions(role, [
+        {
+          id: 'AwsSolutions-IAM5',
+          reason: 'Resource * is needed for CloudWatch Logs policies used on Lambda functions.',
+        },
+      ]);
+
+      const lambdaFunction = new Function(this, 'Function', {
+        role,
+        runtime: props.runtimePython,
+        code: Code.fromBucket(
+          Bucket.fromBucketName(this, 'Bucket', `${props.solutionDistBucket}-${Stack.of(this).region}`),
+          props.solutionTMN + '/' + props.solutionVersion + '/lambda/wait_provider.zip'
+        ),
+        handler: 'wait_provider.lambda_handler',
+        environment: { LOG_LEVEL: 'WARNING' },
+        timeout: Duration.minutes(15),
+      });
+
+      this.serviceToken = lambdaFunction.functionArn;
+    } else {
+      this.serviceToken = props.serviceToken;
+    }
+  }
+
+  private currentWaitResource: CustomResource | undefined;
+  private hash: Hash | undefined;
+  private readonly maxBatchedDocuments = 5;
+  private batchedDocuments = 0;
+
+  createWaitResource(scope: Construct, id: string, props: SsmDocumentWaitResourceProps): CustomResource {
```

Comment:
This is the difference from the non-batched version.

Reply:
is there a way to pass all the SSM documents that need batching at once? The use of a stateful counter between method calls has quite a code smell to it

Reply:
I'm not sure what you mean. There's nothing wrong with a class keeping track of its internal state. The interface is exactly the same as before: when you create an SSM document, call this class to get a resource to wait on. The caller doesn't need to know that that resource is the same for a few other documents, and the callee only needs to maintain the waitResource as needed. If you do it all at once, you do the same thing, just in a loop instead of rolling on each call.

You could do something at the end, but when is that? How do I hook into cdk "when everything else is done, also wrap everything up here"? Is that something that an Aspect is good for? I don't know. Haven't thought too much about the "right" way to do this. I suspect there is none. An additional complication is one that I called out: for the hash to make any sense, we have to batch the same N documents together whenever the solution is built, even as documents are added or removed. That is going to add a fair bit of complexity.